### PR TITLE
Routing

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Complete.

### CODE REVIEW

File: src/OpenLoco/Vehicles/VehicleTail.cpp [hunk 1]
```diff
@@ -211,20 +386,265 @@ namespace OpenLoco::Vehicles
...
+            return;
+        }
+
+        const auto tad = interest.tad();
+        auto nextLoc = interest.loc;
+        if (tad.isReversed())
+        {
+            auto& trackSize = Map::TrackData::getUnkTrack(tad._data);
+            nextLoc += trackSize.pos;
+            if (trackSize.rotationEnd < 12)
```
Comment (duncanspumpkin) [inline src/OpenLoco/Vehicles/VehicleTail.cpp:473]: Confirm if this is rotation end

**Comment by duncanspumpkin** [comment]
PR is now deterministic after 1000 ticks.

File: src/OpenLoco/Vehicles/VehicleTail.cpp [hunk 2]
```diff
@@ -10,9 +12,172 @@ using namespace OpenLoco::Literals;
...
+            bool operator==(Iterator& other) const
+            {
+                return _index == other._index;
+            }
+            bool operator!=(Iterator& other) const
+            {
+                return !(*this == other);
+            }
+
+            std::tuple<Map::Pos2, ZAndTD, CAndT> operator*() const
```
Comment (duncanspumpkin) [inline src/OpenLoco/Vehicles/VehicleTail.cpp:102]: Change this to return a LocationOfInterest

**Comment by duncanspumpkin** [comment]
I've called this function ReduceTransform as it takes a reduction function and a transform function.

**Comment by duncanspumpkin** [comment]
Hmm perhaps filterTransform would be better name.

File: src/OpenLoco/Vehicles/Vehicle.cpp [hunk 3]
```diff
@@ -63,6 +63,18 @@ namespace OpenLoco::Vehicles
         return veh->track_type;
     }
 
+    Map::Pos3 VehicleBase::getTrackLoc() const
+    {
+        const auto* veh = reinterpret_cast<const VehicleCommon*>(this);
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Vehicles/Vehicle.cpp:68]: `VehicleBase` has an `as` member, right? Should that be used instead?

File: src/OpenLoco/Vehicles/Vehicle.h [hunk 4]
```diff
@@ -157,6 +158,10 @@ namespace OpenLoco::Vehicles
     };
     static_assert(sizeof(TrackAndDirection) == 2);
 
+    // TODO move to a different header
```
Comment (LeftofZen) [inline src/OpenLoco/Vehicles/Vehicle.h:161]: is this still todo?

Comment (duncanspumpkin) [reply src/OpenLoco/Vehicles/Vehicle.h:161]: Yes

Comment (duncanspumpkin) [reply src/OpenLoco/Vehicles/Vehicle.h:161]: It will change a good few parts of the code moving this so i don't want to touch it in this pr.

Comment (LeftofZen) [reply src/OpenLoco/Vehicles/Vehicle.h:161]: no worries, just wanted to check

**Review by LeftofZen — APPROVED** [review]
Anecdotally, I tested for a while on multiple save games and saw no difference in vehicle routing behaviour. The code itself is pretty dense for me and I can't be sure it is all perfect, but at a first lookthrough it seems good

**Pull request merged**